javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
   def end(): Int = start() + group().length
   def group(): String = ensureLastMatch(0).get
 
+  lazy val startOf: (/*Group number*/Int => /*Start position*/Int) = {
+    // We wrap every char in the regexp in a group to find the intermediate positions.
+    // Same for subgroups.
+    val regex = pattern0.toString
+    // (?:AB)C => (?:(AB))(?:(C))
+    // Transform:  (A(B))C(D)  to  (((A)(B))(C)(D)) so that using the length of the 1st group and the 3rd, we can obtain the start position of D.
+    
+    /** @param originalGroup Either the 1-based index of the original group that this node encloses, or the modifiers which are present before and after (ex: ?, *?, (?=...), (?!..))
+      * @param 
+      */
+    case class Node(originalGroup: Either[Int, (String, String)], regex: Either[String, Seq[Node]]) {
+      var newGroup: Int = 0
+      var matched: String = ""
+      var start = 0
+      
+      // Takes a new group index and returns a new group index.
+      def setNewGroup(newGroupIndex: Int): Int = {
+        newGroup = newGroupIndex
+        regex match {
+          case Left(r) => newGroupIndex + 1
+          case Right(children) =>
+            ((newGroupIndex + 1) /: children) {
+              case (newgi, c) => c.setNewGroup(newgi)
+            }
+        }
+      }
+    
+      // First we record the matches
+      def setMatch(mr: Int => String): Unit = {
```

Comment:
What does `mr` mean? Consider writing it in full. In general, avoid abbreviations except for codebase-wide well-known conventions (e.g., `pos` for `position`).

Reply:
I renamed these variables and others.